### PR TITLE
Error boundary in extensions

### PR DESCRIPTION
**Pull request by joaomilho**
## Hey, I just made a Pull Request!

> The goal here is to discuss the impl and see if we're too far off

A default ErrorBoundary around the extensions. Assumptions here are that one wants an ErrorBoundary wrapping all plugins touch points (as stated in composabilty doc) with a default design and behavior but with the ability to override its design and behavior per backstage app (installation).

This override is done by defining an ErrorBoundaryFallback component in createApp. This is plugin-aware, since it receives the plugin that generated the error as a prop. This can be used to lookup further info like team name or Slack channel.

We decided not to be opinionated when it comes to this lookup. We also entertained the idea of plugins themselves being able to provide or enhance this default behavior (e.g. a Sentry plugin that automatically reports and returns an errorId), but preferred to stick to the LCM ("least common multiple") that we believe everyone will need and can build on top.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

error-boundary

### CODE REVIEW

File: .changeset/few-cats-type.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+---
+'@backstage/core': minor
+'@backstage/core-api': minor
```
Comment (freben) [inline .changeset/few-cats-type.md:3]: probably make this patch, right?

File: packages/core/src/components/ErrorBoundaryFallback/ErrorBoundaryFallback.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import { ResponseErrorPanel } from '../ResponseErrorPanel';
+import { Button } from '@material-ui/core';
+
+export const ErrorBoundaryFallback: AppComponents['ErrorBoundaryFallback'] = ({
+  error,
+  resetErrorBoundary,
+  plugin,
+}) => {
+  return (
+    <ResponseErrorPanel
```
Comment (freben) [inline packages/core/src/components/ErrorBoundaryFallback/ErrorBoundaryFallback.tsx:30]: Hmm this one was meant specifically for response errors, but... Yeah maybe we are in need of a panel like this that understands all known errors and isn't specific to ResponseError. OK keep this as is and we'll revisit how to help all users display all kinds of errors

File: packages/core-api/src/app/types.ts [hunk 3]
```diff
@@ -59,6 +60,11 @@ export type AppComponents = {
   BootErrorPage: ComponentType<BootErrorPageProps>;
   Progress: ComponentType<{}>;
   Router: ComponentType<{}>;
+  ErrorBoundaryFallback: ComponentType<
+    ErrorBoundaryFallbackProps & {
+      plugin: BackstagePlugin;
```
Comment (freben) [inline packages/core-api/src/app/types.ts:65]: make this optional - we want to be able to place one at the app root as well probably

File: packages/core/src/components/ResponseErrorPanel/ResponseErrorPanel.tsx [hunk 4]
```diff
@@ -107,12 +109,21 @@ const ResponseErrorList = ({
...
+      )}
     </List>
   );
 };
 
 type Props = {
   error: Error;
+  defaultExpanded?: boolean;
+  title?: string;
+  actions?: React.ReactNode;
```
Comment (freben) [inline packages/core/src/components/ResponseErrorPanel/ResponseErrorPanel.tsx:126]: Not a huge fan of stacking more and more props onto this one if we don't have to - long term it leads to complex beahvior and interfaces. Would it be possible to instead render your own `WarningPanel` containing a `ResponseErrorDetails` and then actions below that? Composition is often better

**Comment by freben** [comment]
@Rugvip also linked to this https://www.npmjs.com/package/react-native-error-boundary

That seems to be usable directly. Note also its use of an uppercase component prop

File: packages/core-api/src/extensions/ErrorBoundary.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+
+export type ErrorBoundaryFallbackProps = {
+  error: Error;
+  resetErrorBoundary: () => void;
```
Comment (Rugvip) [inline packages/core-api/src/extensions/ErrorBoundary.tsx:21]: To clarify the link to https://www.npmjs.com/package/react-native-error-boundary I'm thinking we use `resetError` here

File: packages/core-api/src/extensions/ErrorBoundary.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+type FallbackRender = React.FunctionComponent<ErrorBoundaryFallbackProps>;
+
+type Props = {
+  fallbackRender: FallbackRender;
```
Comment (Rugvip) [inline packages/core-api/src/extensions/ErrorBoundary.tsx:27]: And `FallbackComponent` here. Components can be inlined as functions now anyway, so render functions are a bit redundant

File: packages/core-api/src/extensions/ErrorBoundary.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+type State = { error?: Error };
+
+export class ErrorBoundary extends React.Component<Props, State> {
```
[the remark this answers is not in the document]
Comment (joaomilho) [reply packages/core-api/src/extensions/ErrorBoundary.tsx]: Yeap, renamed everything to include the "Plugin" prefix

File: .changeset/few-cats-type.md (hunk 1 above)
Comment (joaomilho) [reply .changeset/few-cats-type.md:3]: k

File: packages/core-api/src/extensions/ErrorBoundary.tsx (hunk 6 above)
Comment (joaomilho) [reply packages/core-api/src/extensions/ErrorBoundary.tsx]: WDYM here?

If the type of this becomes a component, then the component class type fails with "has no call signatures".

Comment (Rugvip) [reply packages/core-api/src/extensions/ErrorBoundary.tsx]: I'm thinking you'd mirror how it's set up here: https://github.com/carloscuesta/react-native-error-boundary/blob/master/src/index.d.ts

**Comment by freben** [comment]
@joaomilho We're moving things around a bit to fit the `core-{app,plugin}-api` structure, fyi

**Review by Rugvip — APPROVED** [review]
🎉 , :shipit:

**Pull request merged**